pandas/core/resample.py

```diff
@@ -387,20 +387,28 @@ def _groupby_and_aggregate(self, how, grouper=None, *args, **kwargs):
 
         obj = self._selected_obj
 
+        result = None
         try:
             grouped = groupby(obj, by=None, grouper=grouper, axis=self.axis)
         except TypeError:
 
             # panel grouper
             grouped = PanelGroupBy(obj, grouper=grouper, axis=self.axis)
-
-        try:
             result = grouped.aggregate(how, *args, **kwargs)
-        except Exception:
 
-            # we have a non-reducing function
-            # try to evaluate
-            result = grouped.apply(how, *args, **kwargs)
+        if result is None:
+            try:
+                if compat.callable(how):
+                    # Check if the function is reducing or not.
+                    result = grouped._aggregate_item_by_item(how,
+                                                             *args, **kwargs)
+                else:
+                    result = grouped.aggregate(how, *args, **kwargs)
+            except Exception:
```

Comment:
can you catch a more specific exception here, maybe TypeError and/or ValueError

Reply:
@jreback 

It catches a lot of exceptions, including base `Exception` as well. I found even
```
self = <pandas.core.groupby.DataFrameGroupBy object at 0x7f7e5a3ce048>, how = 'mean', alt = None, numeric_only = True

    def _cython_agg_blocks(self, how, alt=None, numeric_only=True):
        # TODO: the actual managing of mgr_locs is a PITA
        # here, it should happen via BlockManager.combine
    
        data, agg_axis = self._get_data_to_aggregate()
    
        if numeric_only:
            data = data.get_numeric_data(copy=False)
    
        new_blocks = []
        new_items = []
        deleted_items = []
        for block in data.blocks:
    
            locs = block.mgr_locs.as_array
            try:
                result, _ = self.grouper.aggregate(
                    block.values, how, axis=agg_axis)
            except NotImplementedError:
                # generally if we have numeric_only=False
                # and non-applicable functions
                # try to python agg
    
                if alt is None:
                    # we cannot perform the operation
                    # in an alternate way, exclude the block
                    deleted_items.append(locs)
                    continue
    
                # call our grouper again with only this block
                obj = self.obj[data.items[locs]]
                s = groupby(obj, self.grouper)
                result = s.aggregate(lambda x: alt(x, axis=self.axis))
                newb = result._data.blocks[0]
    
            finally:
    
                # see if we can cast the block back to the original dtype
>               result = block._try_coerce_and_cast_result(result)
E               UnboundLocalError: local variable 'result' referenced before assignment
```

Reply:
that last is quite tricky, this entire routine is actually *also* wrapped in try/except at a higher level to basically do a ``.apply``  (rather than try cython) if things fail. many edge cases here.

you can try o debug that unboundedlocal error, that doesn't seem right (prob result is not defined, and it should raise a sensible error rather than try to coerce a block if all else fails.

Reply:
@jreback 

May I do not touch `except Exception:` line? It seems like a separate issue.

I've found next exceptions there: `IndexError`, `UnboundLocalError`, `AssertionError`, `pandas.core.base.DataError`.